dev/DynamicDependency/MddBootstrap.cpp

```diff
@@ -0,0 +1,378 @@
+void RemoveFrameworkFromPath(PCWSTR frameworkPath);
+CLSID FindDDLM(const PACKAGE_VERSION minVersion);
+CLSID GetClsid(const winrt::Windows::ApplicationModel::AppExtensions::AppExtension appExtension);
+
+IDynamicDependencyLifetimeManager* g_lifetimeManager{};
+wil::unique_hmodule g_projectReunionDll{};
+wil::unique_process_heap_string g_packageDependencyId;
+MDD_PACKAGEDEPENDENCY_CONTEXT g_packageDependencyContext{};
+
+static std::wstring g_test_ddlmPackageNamePrefix;
+static std::wstring g_test_ddlmPackagePublisherId;
+
+inline winrt::Windows::System::ProcessorArchitecture GetCurrentArchitecture()
+{
+#if defined(_M_X64)
+    return winrt::Windows::System::ProcessorArchitecture::X64;
+#elif defined(_M_IX86)
+    return winrt::Windows::System::ProcessorArchitecture::X86;
+#elif defined(_M_ARM64)
+    return winrt::Windows::System::ProcessorArchitecture::Arm64;
+#elif defined(_M_ARM)
+    return winrt::Windows::System::ProcessorArchitecture::Arm;
+#else
+#   error "Unknown processor architecture"
+#endif
+}
+
+inline PCWSTR GetCurrentArchitectureAsString()
+{
+#if defined(_M_X64)
+    return L"x64";
+#elif defined(_M_IX86)
+    return L"x86";
+#elif defined(_M_ARM64)
+    return L"arm64";
+#elif defined(_M_ARM)
+    return L"arm";
+#else
+#   error "Unknown processor architecture"
+#endif
+}
+
+inline winrt::Windows::System::ProcessorArchitecture ParseArchitecture(PCWSTR architecture)
+{
+    if (CompareStringOrdinal(architecture, -1, L"x64", -1, TRUE) == CSTR_EQUAL)
+    {
+        return winrt::Windows::System::ProcessorArchitecture::X64;
+    }
+    else if (CompareStringOrdinal(architecture, -1, L"x86", -1, TRUE) == CSTR_EQUAL)
+    {
+        return winrt::Windows::System::ProcessorArchitecture::X86;
+    }
+    else if (CompareStringOrdinal(architecture, -1, L"arm64", -1, TRUE) == CSTR_EQUAL)
+    {
+    return winrt::Windows::System::ProcessorArchitecture::Arm64;
+    }
+    else if (CompareStringOrdinal(architecture, -1, L"arm", -1, TRUE) == CSTR_EQUAL)
+    {
+    return winrt::Windows::System::ProcessorArchitecture::Arm;
+    }
+    else
+    {
+        return winrt::Windows::System::ProcessorArchitecture::Unknown;
+    }
+}
+
+STDAPI MddBootstrapInitialize(
+    const PACKAGE_VERSION minVersion) noexcept try
+{
+    FAIL_FAST_HR_IF(HRESULT_FROM_WIN32(ERROR_ALREADY_INITIALIZED), g_lifetimeManager != nullptr);
+    FAIL_FAST_HR_IF(HRESULT_FROM_WIN32(ERROR_ALREADY_INITIALIZED), g_projectReunionDll != nullptr);
+    FAIL_FAST_HR_IF(HRESULT_FROM_WIN32(ERROR_ALREADY_INITIALIZED), g_packageDependencyId != nullptr);
+    FAIL_FAST_HR_IF(HRESULT_FROM_WIN32(ERROR_ALREADY_INITIALIZED), g_packageDependencyContext != nullptr);
+
+    const auto appDynamicDependencyLifetimeManagerClsid{ FindDDLM(minVersion) };
+
+    wil::com_ptr_nothrow<IDynamicDependencyLifetimeManager> lifetimeManager(wil::CoCreateInstance<IDynamicDependencyLifetimeManager>(appDynamicDependencyLifetimeManagerClsid, CLSCTX_LOCAL_SERVER));
```

Comment:
A fine place to use `auto`

Reply:
I've been using auto where the type is relatively obvious